### PR TITLE
fix(x86): derive stochastic symbolic registers from target

### LINKED ISSUE #195: x86 stochastic/symbolic should derive register pool from target like enumerative does
Deferred from #184: `run_x86_stochastic` and `run_x86_symbolic` use the global 8-GPR pool from `default_x86_registers()`, but the enumerative path (`find_shorter_equivalent_x86`) builds its pool from registers actually appearing in the target.

Original feedback by @claude bot on PR #184 (https://github.com/pmatos/s11/pull/184#issuecomment-4476263368):

> `src/main.rs`, `run_x86_stochastic` / `run_x86_symbolic` use `default_x86_registers()` unconditionally, whereas the enumerative path builds the pool from registers actually appearing in the target. This isn't a correctness issue, but it unnecessarily widens the x86 search space.

**Context:** `src/main.rs:1056` (run_x86_stochastic), `src/main.rs:1109` (run_x86_symbolic). Compare with `find_shorter_equivalent_x86` in the same file.

**Why deferred:** scope-creep — search-space optimization, not correctness. The plan's Phase E explicitly uses `default_x86_registers()`. A unified register-pool-construction policy across enumerative / stochastic / symbolic is its own design discussion.

_Filed automatically by `pm-autofix-pr` after dual-evaluator triage by Claude Opus 4.7 and Codex GPT-5.5._

### PR DESCRIPTION
**Pull request by pmatos**
Summary:
- share the x86 target-derived register pool policy across enumerative, stochastic, and symbolic config builders
- filter RSP and RBP out of target-derived x86 search register pools so candidates cannot make stack/frame registers writable
- keep the default register fallback only for empty targets; non-empty all stack/frame targets now produce an empty pool instead of unrelated writable defaults
- make x86 stochastic search skip when the filtered pool is empty, avoiding backend fallback registers
- keep stochastic/symbolic immediate pools at their existing defaults
- add CLI helper regression tests for first-seen register derivation, stack/frame exclusion, empty-pool behavior, and x86 stochastic/symbolic config preservation

Closes #195

**Merge with `gh pr merge --merge` (merge commit, not squash) per CLAUDE.md.**

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -917,7 +921,26 @@ fn build_x86_stochastic_search_config(width: u32, options: &OptimizationOptions)
...
+    let symbolic_config = SymbolicConfig::default()
+        .with_search_mode(options.search_mode)
+        .with_timeout(options.solver_timeout);
+
+    SearchConfig::default()
+        .with_symbolic(symbolic_config)
+        .with_cost_metric(options.cost_metric)
+        .with_timeout_option(options.timeout)
+        .with_verbose(options.verbose)
+        .with_x86_registers(x86_registers_from_target(target))
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/main.rs:943]: Fixed in 0e56a4c by filtering `X86Register::RSP` in the shared `x86_registers_from_target` helper before stochastic/symbolic configs receive the candidate pool. The fallback still returns `default_x86_registers()` when the target has no usable non-RSP registers, so all-RSP targets do not produce an empty generator pool.

Covered by:
- `cargo test --bin s11 cli_helper_tests::x86_register_pool_is_first_seen_target_derived_and_empty_falls_back`
- `cargo test --bin s11 cli_helper_tests::build_x86_stochastic_search_config_propagates_solver_timeout`
- `cargo test --bin s11 cli_helper_tests::build_x86_symbolic_search_config_is_target_derived_and_preserves_symbolic_options`

File: src/main.rs [hunk 2]
```diff
@@ -1662,6 +1686,9 @@ fn x86_enumerative_registers_from_target(
             .chain(instr.source_registers())
     });
     for reg in referenced {
+        if reg == isa::x86::X86Register::RSP {
+            continue;
+        }
```
[the remark this answers is not in the document]
Comment (pmatos) [reply src/main.rs]: Fixed in cc4bc47 by taking the writable-pool option rather than expanding live-out: `x86_registers_from_target` now filters both `RSP` and `RBP`, and the default register fallback is limited to the truly empty target. Non-empty targets whose only referenced registers are stack/frame registers now produce an empty pool instead of default writable registers.

I also added an x86 stochastic guard for an empty filtered pool so it returns no optimization instead of reaching the backend random-register fallback. Symbolic/enumerative receive an empty pool and therefore enumerate no writable candidates.

Covered by the focused CLI helper tests plus the final `./ci_check.sh` run.

File: src/main.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (pmatos) [reply src/main.rs:943]: Fixed in 313bcbf by taking the writable-pool option from the review: the shared x86 search register helper now derives candidate registers from original destinations only, still filtering RSP/RBP and still falling back to defaults only for an actually empty target. Source-only registers such as R10/R11 no longer reach enumerative/stochastic/symbolic writable pools.

I also removed the stochastic x86 empty-pool RAX fallback so an intentionally empty writable pool does not get widened again during random sequence generation or instruction mutation.

Covered by:
- `cargo test --bin s11 cli_helper_tests::x86_register_pool_is_destination_derived_and_empty_falls_back`
- `cargo test --bin s11 cli_helper_tests::build_x86`
- `cargo test empty_register_pool`
- `cargo test x86`
- `cargo clippy --all -- -D warnings`
- `cargo test --all`
- `./ci_check.sh`

**Comment by pmatos** [comment]
**Triage of the automated review on `3f0e3123`** — each suggestion was evaluated by two independent AI reviewers (Claude Opus 4.8 and Codex GPT-5.5):

- **Enumerative delegates to the stochastic builder** — **No action taken.** `build_x86_enumerative_search_config` already carries a docstring explaining the intentional target-derived / immediates / cores relationship; the inherited `StochasticConfig` fields are unused by the enumerative path and harmless. The deeper *refactor* (extract a shared base builder) is legitimate but out of scope → **deferred**, tracked as #457.
- **Add a comment noting the parallel stochastic/symbolic test targets** — **Style preference (no change).** The two config tests are self-explanatory; the annotation is cosmetic.
- **No end-to-end R10–R15 rewrite test for stochastic/symbolic** — **Deferred (diminishing returns).** The config-level tests already pin the register-pool fix; an end-to-end stochastic/symbolic test adds Z3 + assembly round-trip cost and flakiness for an optional non-blocker. Tracked as #458.

_Outcomes above were agreed by both reviewers. If you disagree, reply and we'll re-evaluate._

**Comment by pmatos** [comment]
**Triage of the automated review on `0e56a4cc`** — evaluated by two independent AI reviewers (Claude Opus 4.8 and Codex GPT-5.5):

- **PR description should note enumerative also gets RSP exclusion** — **Already handled (no change).** The PR body already states "filter RSP and RBP out of target-derived x86 search register pools" and "share the x86 target-derived register pool policy across enumerative, stochastic, and symbolic config builders" — the enumerative path is explicitly covered.
- **No enumerative RSP/RBP regression test** — **Already handled (no change).** Enumerative routes through the shared `x86_registers_from_target` helper, whose RSP/RBP (and source-only) exclusion is directly asserted by `x86_register_pool_is_destination_derived_and_empty_falls_back`; the enumerative config test additionally asserts R10/RAX exclusion.
- **Comment on the inherited unused stochastic fields / extract a shared base builder** — **No change** for the inline comment (the docstring already covers intent); the refactor is **deferred**, tracked as #457.

_Outcomes above were agreed by both reviewers. If you disagree, reply and we'll re-evaluate._

**Comment by pmatos** [comment]
**Triage of the automated review on `cc4bc47b`** — evaluated by two independent AI reviewers (Claude Opus 4.8 and Codex GPT-5.5):

- **Missing empty-pool guard in `run_x86_symbolic`** — **Not an issue (no change).** Verified that an empty `x86_available_registers` pool cannot panic or misbehave in the symbolic path: `X86InstructionGenerator::generate_all` builds candidates only inside `for &rd in registers { ... }` loops, so an empty pool yields an empty candidate list; `linear_search` / `search_at_length` then iterate `for instr in all_instructions` (a no-op) and return `None`. No slice indexing is reached.

  The stochastic guard exists for a different, backend-specific reason: the stochastic path uses random proposal/mutation (which index into the register slice) and would otherwise run useless work over an intentionally-empty pool. The symbolic path enumerates from the exact pool and has no random/fallback widening, so the asymmetry is intentional and safe.

_This assessment was made by both reviewers. If you disagree, reply and we'll re-evaluate._

**Comment by pmatos** [comment]
**Triage of the automated review on `313bcbfe`** — evaluated by two independent AI reviewers (Claude Opus 4.8 and Codex GPT-5.5):

- **Verify no other `mutate_opcode` opcode-bridge calls `pick_register` without handling `None`** — **Already handled (no change).** Confirmed: in x86 `mutate_opcode` the only `pick_register` call is the `CmpImm → CmpReg` bridge, which already returns the unchanged instruction on `None`; the sibling `CmpReg → CmpImm` arm uses the infallible `pick_immediate`. The `.expect("register pool is non-empty")` calls live in `mutate_operand`, which is guarded by the empty-pool early return.
- **Enumerative delegates to the stochastic builder / inherited unused fields** — **No change** (the docstring covers intent); the base-builder refactor is **deferred**, tracked as #457.
- **Comment noting the parallel test targets** — **Style preference (no change).**

_Outcomes above were agreed by both reviewers. If you disagree, reply and we'll re-evaluate._
